File: python/context_engineering/compaction.py
"""Compaction module: automatic context management for multi-turn agents.

Tracks token budgets across turns and compacts old content to stay in budget.
Think of it as "malloc for context windows."
"""

from __future__ import annotations

import time
from dataclasses import dataclass
from typing import Any, Callable, Dict, List, Optional

from .core import Budget, ContextItem, create_causal_scorer, estimate_tokens
from .errors import ValidationError


@dataclass
class Turn:
    """A conversation turn."""

    role: str
    content: str
    tokens: int = 0
    timestamp: float = 0.0
    is_summary: bool = False
    task_id: Optional[str] = None
    is_outcome: Optional[bool] = None


@dataclass
class CompileResult:
    """Result of compiling context."""

    turns: List[Turn]
    items: List[ContextItem]
    total_tokens: int


class ContextManager:
    """Manages context across turns with automatic compaction.

    Tracks token budgets, preserves recent turns verbatim, and compacts
    older turns into summaries when they exceed the summarize threshold.
    """

    def __init__(
        self,
        budget: Budget,
        summarize_after_turns: int = 5,
        preserve_recent_turns: int = 2,
        system_prompt: Optional[str] = None,
        token_estimator: Optional[Callable[[str], int]] = None,
    ) -> None:
        self._budget = budget
        self._summarize_after = summarize_after_turns
        self._preserve_recent = preserve_recent_turns
        self._system_prompt = system_prompt
        self._estimate = token_estimator or (lambda text: estimate_tokens(text))
        self._turns: List[Turn] = []
        self._items: List[ContextItem] = []
        self._active_task_id: Optional[str] = None
        self._beads_graph: List[Any] = []

        self._system_tokens = self._estimate(system_prompt) if system_prompt else 0
        self._effective_budget = budget.max_tokens - (budget.reserve_tokens or 0)

    def set_active_task(self, task_id: str) -> None:
        """Set the currently active task ID."""
        self._active_task_id = task_id

    def set_beads_graph(self, issues: List[Any]) -> None:
        """Provide a BEADS graph for causal scoring."""
        self._beads_graph = issues

    def add_turn(
        self, role: str, content: str, task_id: Optional[str] = None, is_outcome: bool = False
    ) -> None:
        """Add a conversation turn."""
        tokens = self._estimate(content)
        tid = task_id or self._active_task_id
        self._turns.append(
            Turn(
                role=role,
                content=content,
                tokens=tokens,
                timestamp=time.time(),
                task_id=tid,
                is_outcome=is_outcome,
            )
        )

    def add_items(self, items: List[ContextItem]) -> None:
        """Add context items (e.g., from memory queries)."""
        self._items.extend(items)

    def get_token_usage(self) -> Dict[str, int]:
        """Get current token usage breakdown."""
        turn_tokens = sum(t.tokens for t in self._turns)
        item_tokens = sum(i.tokens or self._estimate(i.content) for i in self._items)
        used = self._system_tokens + turn_tokens + item_tokens
        return {
            "used": used,
            "budget": self._effective_budget,
            "remaining": max(0, self._effective_budget - used),
        }

    def compile(self) -> CompileResult:
        """Compile context -- returns turns + items that fit within budget.

        Three phases:
        1. Preserve recent turns verbatim
        2. Compact older turns (causal scoring if graph available, else summary)
        3. Pack context items into remaining budget
        """
        available = self._effective_budget - self._system_tokens

        # Phase 1: Preserve recent turns
        if self._preserve_recent > 0:
            recent_turns = self._turns[-self._preserve_recent :]
            older_turns = self._turns[: -self._preserve_recent]
        else:
            recent_turns = []
            older_turns = list(self._turns)

        recent_tokens = sum(t.tokens for t in recent_turns)
        available -= recent_tokens

        # Phase 2: Compact older turns
        compacted_older: List[Turn] = []

        # If we have a BEADS graph, we use causal scoring
        scorer = None
        if self._beads_graph:
            scorer = create_causal_scorer(self._beads_graph, self._active_task_id)

        if scorer and older_turns:
            # Map Turns to ContextItems for scoring
            scored_turns = []
            for idx, t in enumerate(older_turns):
                item = ContextItem(
                    id=f"turn-{idx}",
                    content=t.content,
                    tokens=t.tokens,
                    task_id=t.task_id,
                    is_outcome=t.is_outcome,
                    priority=5.0,
                    recency=t.timestamp,
                )
                score = scorer(item)
                scored_turns.append((t, score))

            # Sort by causal score
            scored_turns.sort(key=lambda pair: pair[1], reverse=True)

            for t, _ in scored_turns:
                if t.tokens <= available:
                    compacted_older.append(t)
                    available -= t.tokens

            # Re-sort by timestamp for conversation order
            compacted_older.sort(key=lambda t: t.timestamp)

        elif older_turns and len(older_turns) >= self._summarize_after:
            combined = "\n".join(f"[{t.role}]: {t.content}" for t in older_turns)
            target_tokens = int(available * 0.3)
            # Binary search for the right truncation point to hit target_tokens.
            # Start with a heuristic then adjust.
            lo, hi = 0, len(combined)
            truncated = combined
            while lo < hi:
                mid = (lo + hi) // 2
                candidate = combined[:mid]
                est = self._estimate(candidate)
                if est <= target_tokens:
                    truncated = candidate
                    lo = mid + 1
                else:
                    hi = mid
            summary_tokens = self._estimate(truncated)

            compacted_older.append(
                Turn(
                    role="system",
                    content=f"[Summary of {len(older_turns)} earlier turns]\n{truncated}",
                    tokens=summary_tokens,
                    is_summary=True,
                    timestamp=older_turns[0].timestamp if older_turns else 0.0,
                )
            )
            available -= summary_tokens
        else:
            for turn in older_turns:
                if turn.tokens <= available:
                    compacted_older.append(turn)
                    available -= turn.tokens

        # Phase 3: Pack context items into remaining budget
        selected_items: List[ContextItem] = []
        if self._items and available > 0:
            item_scorer = scorer if scorer else (lambda i: i.score or 0.0)

            scored = []
            for item in self._items:
                tokens = item.tokens or self._estimate(item.content)
                item_with_tokens = item.model_copy(update={"tokens": tokens})
                score = item_scorer(item_with_tokens)
                scored.append((item_with_tokens, score))

            scored.sort(key=lambda pair: pair[1], reverse=True)

            used_item_tokens = 0
            for item, _ in scored:
                if used_item_tokens + (item.tokens or 0) <= available:
                    selected_items.append(item)
                    used_item_tokens += item.tokens or 0

        all_turns = compacted_older + recent_turns
        total_tokens = (
            self._system_tokens
            + sum(t.tokens for t in all_turns)
            + sum(i.tokens or self._estimate(i.content) for i in selected_items)
        )

        return CompileResult(
            turns=all_turns,
            items=selected_items,
            total_tokens=total_tokens,
        )

    def turn_count(self) -> int:
        """Get the number of turns."""
        return len(self._turns)

    def clear(self) -> None:
        """Clear all turns and items."""
        self._turns = []
        self._items = []
        self._active_task_id = None
        self._beads_graph = []


def create_context_manager(
    budget: Budget,
    summarize_after_turns: int = 5,
    preserve_recent_turns: int = 2,
    system_prompt: Optional[str] = None,
    token_estimator: Optional[Callable[[str], int]] = None,
) -> ContextManager:
    """Create an automatic context compaction manager.

    Args:
        budget: Token budget with max_tokens and optional reserve_tokens.
        summarize_after_turns: Compact older turns after this many (default: 5).
        preserve_recent_turns: Always keep last N turns verbatim (default: 2).
        system_prompt: System prompt to always include.
        token_estimator: Custom token estimator function.

    Returns:
        A ContextManager instance.
    """
    if budget.max_tokens <= 0:
        raise ValidationError(
            "Invalid budget",
            details=[
                {
                    "path": "budget.max_tokens",
                    "message": "max_tokens must be greater than 0",
                }
            ],
        )

    return ContextManager(
        budget=budget,
        summarize_after_turns=summarize_after_turns,
        preserve_recent_turns=preserve_recent_turns,
        system_prompt=system_prompt,
        token_estimator=token_estimator,
    )

File: packages/ce-core/src/compaction.ts
import type {
  Budget,
  ContextItem,
  Summarizer,
  TokenEstimator,
} from "./types.js";
import type { BeadsIssue } from "./beads.js";
import { defaultTokenEstimator } from "./estimate.js";
import { createCausalScorer } from "./score.js";

export interface Turn {
  role: "user" | "assistant" | "tool" | "system";
  content: string;
  tokens?: number;
  timestamp?: number;
  /** If true, this turn is a summary of older turns */
  isSummary?: boolean;
  /** The BEADS task ID this turn belongs to */
  taskId?: string;
  /** If true, this turn represents a critical outcome/result of a task */
  isOutcome?: boolean;
}

export interface CompactionOptions {
  /** Maximum token budget for the managed context */
  budget: Budget;
  /** Summarize tool results older than N turns (default: 5) */
  summarizeAfterTurns?: number;
  /** Always preserve the last N turns verbatim (default: 2) */
  preserveRecentTurns?: number;
  /** System prompt to always include */
  systemPrompt?: string;
  /** Token estimator */
  tokenEstimator?: TokenEstimator;
  /** Summarizer for compacting old turns (if not provided, truncation is used) */
  summarizer?: Summarizer;
}

/**
 * A context manager that tracks token budgets across turns
 * and automatically compacts old content to stay within budget.
 *
 * @example
 * ```ts
 * const ctx = createContextManager({
 *   budget: { maxTokens: 8000 },
 *   summarizeAfterTurns: 3,
 *   preserveRecentTurns: 2,
 *   systemPrompt: "You are a helpful assistant.",
 * });
 *
 * ctx.addTurn({ role: "user", content: "What is context engineering?" });
 * ctx.addTurn({ role: "assistant", content: "Context engineering is..." });
 * ctx.addTurn({ role: "tool", content: longSearchResult });
 *
 * const compiled = ctx.compile();
 * // Returns turns that fit within budget, with old tool results summarized
 * ```
 */
export interface ContextManager {
  /** Set the currently active task ID */
  setActiveTask(taskId: string): void;
  /** Provide a BEADS graph for causal scoring */
  setBeadsGraph(issues: BeadsIssue[]): void;
  /** Add a turn to the conversation */
  addTurn(turn: Omit<Turn, "tokens" | "timestamp">): void;
  /** Add context items (e.g., from memory queries) */
  addItems(items: ContextItem[]): void;
  /** Get current token usage */
  getTokenUsage(): { used: number; budget: number; remaining: number };
  /** Compile the context — returns turns + items that fit within budget */
  compile(): { turns: Turn[]; items: ContextItem[]; totalTokens: number };
  /** Get the number of turns */
  turnCount(): number;
  /** Clear all turns and items */
  clear(): void;
}

/**
 * Create a context manager that tracks token budgets and compacts old turns.
 *
 * @param options - Compaction configuration including budget, summarization thresholds, and system prompt
 * @returns A ContextManager instance
 */
export function createContextManager(
  options: CompactionOptions
): ContextManager {
  const estimate = options.tokenEstimator ?? defaultTokenEstimator;
  const summarizeAfter = options.summarizeAfterTurns ?? 5;
  const preserveRecent = options.preserveRecentTurns ?? 2;
  const maxTokens = options.budget.maxTokens;
  const reserveTokens = options.budget.reserveTokens ?? 0;
  const effectiveBudget = maxTokens - reserveTokens;

  let turns: Turn[] = [];
  let items: ContextItem[] = [];
  let activeTaskId: string | undefined;
  let beadsGraph: BeadsIssue[] = [];

  const systemTokens = options.systemPrompt
    ? estimate(options.systemPrompt)
    : 0;

  function setActiveTask(taskId: string): void {
    activeTaskId = taskId;
  }

  function setBeadsGraph(issues: BeadsIssue[]): void {
    beadsGraph = issues;
  }

  function addTurn(turn: Omit<Turn, "tokens" | "timestamp">): void {
    const tokens = estimate(turn.content);
    // Inherit activeTaskId if not provided
    const taskId = (turn as { taskId?: string }).taskId ?? activeTaskId;
    turns.push({ ...turn, taskId, tokens, timestamp: Date.now() });
  }

  function addItems(newItems: ContextItem[]): void {
    items = [...items, ...newItems];
  }

  function getTokenUsage(): {
    used: number;
    budget: number;
    remaining: number;
  } {
    const used =
      systemTokens +
      turns.reduce((sum, t) => sum + (t.tokens ?? 0), 0) +
      items.reduce((sum, i) => sum + (i.tokens ?? estimate(i.content)), 0);
    return {
      used,
      budget: effectiveBudget,
      remaining: Math.max(0, effectiveBudget - used),
    };
  }

  function compile(): {
    turns: Turn[];
    items: ContextItem[];
    totalTokens: number;
  } {
    let availableTokens = effectiveBudget - systemTokens;

    // Phase 1: Always preserve recent turns
    const recentTurns = preserveRecent > 0 ? turns.slice(-preserveRecent) : [];
    const olderTurns =
      preserveRecent > 0 ? turns.slice(0, -preserveRecent) : turns;

    const recentTokens = recentTurns.reduce(
      (sum, t) => sum + (t.tokens ?? 0),
      0
    );
    availableTokens -= recentTokens;

    // Phase 2: Compact older turns
    // If we have a BEADS graph, we use causal scoring to prune older turns
    const scorer = beadsGraph.length > 0 
      ? createCausalScorer(beadsGraph, activeTaskId)
      : undefined;

    const compactedOlder: Turn[] = [];

    if (scorer && olderTurns.length > 0) {
      // Map Turns to ContextItems for scoring
      const scoredTurns = olderTurns.map((t, idx) => {
        const item: ContextItem = {
          id: `turn-${idx}`,
          content: t.content,
          tokens: t.tokens,
          taskId: t.taskId,
          isOutcome: t.isOutcome,
          priority: 5,
          recency: t.timestamp
        };
        return { turn: t, score: scorer(item) };
      });

      // Sort by causal score
      scoredTurns.sort((a, b) => b.score - a.score);

      for (const st of scoredTurns) {
        if ((st.turn.tokens ?? 0) <= availableTokens) {
          compactedOlder.push(st.turn);
          availableTokens -= st.turn.tokens ?? 0;
        }
      }
      // Re-sort by timestamp for conversation order
      compactedOlder.sort((a, b) => (a.timestamp ?? 0) - (b.timestamp ?? 0));

    } else if (olderTurns.length > 0 && olderTurns.length >= summarizeAfter) {
      // Fallback: Group older turns and create a summary
      const combinedContent = olderTurns
        .map(t => `[${t.role}]: ${t.content}`)
        .join("\n");

      const targetTokens = Math.floor(availableTokens * 0.3);
      const truncated = combinedContent.slice(0, targetTokens * 4);
      const summaryTokens = estimate(truncated);

      compactedOlder.push({
        role: "system",
        content: `[Summary of ${olderTurns.length} earlier turns]\n${truncated}`,
        tokens: summaryTokens,
        isSummary: true,
        timestamp: olderTurns[0]?.timestamp,
      });
      availableTokens -= summaryTokens;
    } else {
      // Include older turns as-is
      for (const turn of olderTurns) {
        if ((turn.tokens ?? 0) <= availableTokens) {
          compactedOlder.push(turn);
          availableTokens -= turn.tokens ?? 0;
        }
      }
    }

    // Phase 3: Pack context items into remaining budget
    let selectedItems: ContextItem[] = [];
    if (items.length > 0 && availableTokens > 0) {
      const itemScorer = scorer ?? ((i: ContextItem) => i.score ?? 0);
      
      selectedItems = items
        .map(i => ({ 
          ...i, 
          tokens: i.tokens ?? estimate(i.content)
        }))
        .map(i => ({ ...i, score: itemScorer(i) }))
        .filter(i => (i.tokens ?? 0) <= availableTokens)
        .sort((a, b) => (b.score ?? 0) - (a.score ?? 0));

      let usedTokens = 0;
      selectedItems = selectedItems.filter(item => {
        if (usedTokens + (item.tokens ?? 0) <= availableTokens) {
          usedTokens += item.tokens ?? 0;
          return true;
        }
        return false;
      });
    }

    const allTurns = [...compactedOlder, ...recentTurns];
    const totalTokens =
      systemTokens +
      allTurns.reduce((sum, t) => sum + (t.tokens ?? 0), 0) +
      selectedItems.reduce((sum, i) => sum + (i.tokens ?? 0), 0);

    return { turns: allTurns, items: selectedItems, totalTokens };
  }

  function turnCount(): number {
    return turns.length;
  }

  function clear(): void {
    turns = [];
    items = [];
    activeTaskId = undefined;
    beadsGraph = [];
  }

  return { 
    setActiveTask, 
    setBeadsGraph, 
    addTurn, 
    addItems, 
    getTokenUsage, 
    compile, 
    turnCount, 
    clear 
  };
}
